### PR TITLE
FEAT: New Plugin for pyAV

### PR DESCRIPTION
**Pull request by FirefoxMetzger**
Closes: #610, https://github.com/imageio/imageio/issues/220, https://github.com/imageio/imageio/issues/686

This PR adds a pyAV plugin to ImageIO using the v3 API. This is in parallel to our imageio-ffmpeg plugin, because (a) it doesn't have feature parity and (b) I still need to work out how well maintained pyAV is. One of the reasons why we still want to have a pyAV plugin is because pyAV is more performant and - depending on the task - I get a 3x -10x speedup by switching to this plugin.

Other features:
- sophisticated way of checking if ffmpeg can handle the ImageResource that is based on ffmpeg's logic instead of us manually checking extensions
- ImageProperties (shape+dtype) without reading pixel data 🎉 
- supports reading and writing of videos
- adjustable thread count when reading/decoding
- optimised threading model for bulk-reading (`index=None`)
- custom conversion between av.VideoFrame and np.ndarray to support formats that pyAV doesn't support yet (I think I will try to get this merged upstream at some point)
- support for filters and graphs (arbitrary ones when iterating, zero-delay ones writing writing or reading single frames)
- typing support

Planned features for upcoming PRs: 
- support for streams (needs rework of `core.Requests` to accept relevant schemes)
- Fast-path based on file extension. [This will happen in a separate PR because it is quite substantial (more than 150 supported formats) and I want to keep this PR focussed on code].
- support for libavdevice (again to keep this PR on core functionality)

### CODE REVIEW

**Comment by FirefoxMetzger** [comment]
@hmaarrfk I saw your name pop up while reading through the pyAV issues. Do you have any interest in a plugin like this?

**Comment by hmaarrfk** [comment]
(between .3x - 10x times faster in my naive timings).

Did you forget a 1 somewhere?

**Comment by FirefoxMetzger** [comment]
> (between .3x - 10x times faster in my naive timings).
> 
> Did you forget a 1 somewhere?

True. it's 1.3x - 10x as fast. 1.3x (30% faster) for bulk reading of an entire video (where decoding dominates runtime), about 5x faster for first frame access (where setup cost dominates), and 10x faster for a random frame (where seeking dominates). Although the seeking depends heavily on the contents of the file and how keyframes are distributed, so that number should be taken with some salt.

**Comment by hmaarrfk** [comment]
For what its worth, I've been trying to get the hardware decoders in reasonable versions of ffmpeg to work.

I think I got the intel ones in conda-forge.

I tried to get nvidia ones in, but I think there are some intriciacies with dependencies and driver versions.

So i am generally interested in speedups :D but not sure how to help at this point. I might be able to give a review if that is what you want?

**Comment by hmaarrfk** [comment]
Is there a way to hide codecov from the PR? it is really annoyign to read it.

**Comment by FirefoxMetzger** [comment]
> but not sure how to help at this point. I might be able to give a review if that is what you want?

Sure that would help :) Though, I wouldn't want to take up too much of your time at this stage. Maybe I can ping you once the PR is RTR (so that, among other things, the codecov messages are gone ;) ).

I primarily pinged you for awareness, because I thought you could be interested in a feature like this. 

I haven't yet figured out a good way to disseminate news around ImageIO nor a good way to estimate interest in new/upcoming features, so this is also partly me figuring out a good way to go about this. Ideas are welcome :D

> I tried to get nvidia ones in, but I think there are some intriciacies with dependencies and driver versions.

Interesting. I've always thought of hardware acceleration (especially via GPU) as something that encoding benefits from, but decoding not so much. Primarily, because sending data back and forth isn't exactly cheap. Do you know of a benchmark for this? (I did some searching yesterday, but didn't find any comparisons beyond people guesstimating)

On that note, you are probably aware of [this note](https://pyav.org/docs/develop/overview/about.html?highlight=hardware#unsupported-features) in pyAV that hardware acceleration won't be supported. Not sure how much we can do from here without adding new features to pyAV (e.g., to expose the required config).

**Comment by hmaarrfk** [comment]
`-hwaccel` is one thing, but there are different encoders that can help speed things up.

Thanks for looking into this!

**Comment by FirefoxMetzger** [comment]
> -hwaccel is one thing, but there are different encoders that can help speed things up.

@hmaarrfk Ah, so you are mainly interested in encoding/writing of video?

**Comment by hmaarrfk** [comment]
Decoding too! It is very useful when dealing with large stacks of images!

**Comment by FirefoxMetzger** [comment]
I am done with the bulk of the work for this. I still need to write some narrative docs and fix pypy (though I think it breaks because of the new astropy and not because of pyav), but the code is there and it is covered by tests.

@hmaarrfk Would you have the time in the coming days to review the code section of this PR (the file pyav.py)?

File: tests/test_pyav.py
Comment (hmaarrfk) [inline tests/test_pyav.py:103]: I'm curious why these are fake formats. There are presumably many fake formats. Why choose these?

File: tests/test_pyav.py
Comment (hmaarrfk) [inline tests/test_pyav.py:146]: Do images really come out as `time, color, height, width`? Typically you would think of them as coming as `time, height, width, color`

File: tests/test_pyav.py
Comment (hmaarrfk) [inline tests/test_pyav.py:176]: By default, pyav returns images in `yuv444p`?  Shouldn't you want things as `rgb` by default in a `python` world?

File: imageio/plugins/pyav.py
Comment (hmaarrfk) [inline imageio/plugins/pyav.py:17]: The typing module is really new. Do you need to bump numpy quite high?

**Review by hmaarrfk — COMMENTED** [review]
Just a few comments for now.

File: tests/test_pyav.py
Comment (FirefoxMetzger) [reply tests/test_pyav.py:103]: Maybe `fake_formats` is not a good name for them. All of these make sense, and they are typically related to hardware acceleration (which is currently unsupported in pyAV).

The reason I call them fake formats here is because we can instantiate them as a format (`av.VideoFormat(fake_format_name)` succeeds), but these formats have no components, so we can't deduce an appropriate dtype from these so we have to raise instead.

File: tests/test_pyav.py
Comment (FirefoxMetzger) [reply tests/test_pyav.py:146]: @hmaarrfk Yes, because we don't modify the raw pixel format unless told to do so. 

`cockatoo.mp4` uses `yuv444p`, which is a planar format and it will thus come out as `time, channel, height, width`. Though, we could easily make it something more familiar by passing `format="rgb24"` which would then produce `time, height, width, channel`.

File: tests/test_pyav.py
Comment (FirefoxMetzger) [reply tests/test_pyav.py:176]: @hmaarrfk By default, it returns the pixels in the same format as they were stored in the file. If we don't know how to return that format (e.g. because it's not trivally stridable), we raise an exception and ask the user to explicitly specify a suitable format via the `format=` kwarg, e.g., `format="rgb24"` to return RGB frames.

File: imageio/plugins/pyav.py
Comment (FirefoxMetzger) [reply imageio/plugins/pyav.py:17]: Yes, `ArrayLike` was introduced in numpy 1.20 and we depend on that since ImageIO v2.15.

File: imageio/plugins/pyav.py
Comment (FirefoxMetzger) [inline imageio/plugins/pyav.py:60]:
```suggestion
```

This is no longer the case.

File: tests/test_pyav.py
Comment (FirefoxMetzger) [reply tests/test_pyav.py:146]: @hmaarrfk Yes, because we don't modify the raw pixel format unless told to do so. 

`cockatoo.mp4` uses `yuv444p`, which is a planar format and it will thus come out as `time, channel, height, width`. Though, we could easily make it something more familiar by passing `format="rgb24"` which would then produce `time, height, width, channel`.

Comment (hmaarrfk) [inline tests/test_pyav.py:146]:
```suggestion
    # cockatoo.mp4 uses yuv444p, which is a planar format 
    # and it will thus come out as time, channel, height, width. 
    assert frames.shape == (283, 3, 480, 640)

    frames = iio.imread(
        test_images / "cockatoo.mp4",
        index=None,
        plugin="pyav",
        # passing format "rgb24" ensures the images come out as 
        # time, height, width, channel
        format="rgb24",
        filter_sequence=[
            ("drawgrid", "w=iw/3:h=ih/3:t=2:c=white@0.5"),
            ("scale", {"size": "vga", "flags": "lanczos"}),
            ("tpad", "start=3"),
        ],
    )

    assert frames.shape == (283, 480, 640, 3)
```

File: tests/test_pyav.py
Comment (hmaarrfk) [reply tests/test_pyav.py:176]: understood. thank you.

File: tests/test_pyav.py
Comment (hmaarrfk) [reply tests/test_pyav.py:103]: could you add a comment in the code to that effect
```suggestion
    # All of these make sense, and they are typically related to hardware acceleration,
    # which is  unsupported in pyAV at the time of writing (2022/03/06)

    # We can instantiate them as a format (av.VideoFormat(fake_format_name) succeeds)
    # but these formats have no components. Therefore we can't deduce an appropriate 
    # dtype from these so we have to raise instead.
    fake_formats = [
```

File: tests/test_pyav.py
Comment (almarklein) [reply tests/test_pyav.py:146]: Regardless, the final image provided to the user would be `height, width, color`, right?

Comment (almarklein) [reply tests/test_pyav.py:146]: Oh, it does not. IMO imageio should make sure that it is.

File: imageio/plugins/pyav.py
Comment (almarklein) [inline imageio/plugins/pyav.py:157]: Can we avoid opening twice?

File: imageio/plugins/pyav.py
Comment (almarklein) [inline imageio/plugins/pyav.py:206]:
```suggestion
            faster seeking inside the file. If False, the video is reset before
```

Comment (almarklein) [reply imageio/plugins/pyav.py:206]: Since this can also be auto-detected, what is the need for making this settable with an arg?

File: imageio/plugins/pyav.py
Comment (almarklein) [inline imageio/plugins/pyav.py:236]:
```suggestion
        bidirectionaly predicted pictures. I don't have any test videos of this
```

File: imageio/plugins/pyav.py
Comment (almarklein) [inline imageio/plugins/pyav.py:793]: I always find it helpful if names that represent flags start with `is_`, e.g.  `is_interlaced` and `is_key_frame`. What do you think?

File: imageio/plugins/pyav.py
Comment (almarklein) [inline imageio/plugins/pyav.py:90]: I'm a bit confused. This suggests that its always `height, width, channels`, but one of the tests seems to suggest it does not. I'm probably missing something.

**Review by almarklein — COMMENTED** [review]
Nice!

File: tests/test_pyav.py
Comment (FirefoxMetzger) [reply tests/test_pyav.py:146]: > Oh, it does not. IMO imageio should make sure that it is.

Yes, `p` indicates planar (channel-first) formats. Well technically not channel-first, because you have weird formats like `nv24` which are planar and mix channel-first and channel-last, but yes. Usually, it means channel-first :D

Regarding the discussion of making sure that frames will be channel-last, I think we should continue that discussion in one place (here: https://github.com/imageio/imageio/issues/504)

File: imageio/plugins/pyav.py
Comment (FirefoxMetzger) [reply imageio/plugins/pyav.py:17]: I'm actually now debating if it would make sense to move all the type annotations into a separate repo and to relax the numpy dependency. Not 100% sure if we depend on anything else from `numpy>=1.20`...

File: imageio/plugins/pyav.py
Comment (FirefoxMetzger) [reply imageio/plugins/pyav.py:90]: Hm, maybe the name `channel` is a bit misleading, but I'm not 100% sure what good taxonomy for this should be. 

FFMpeg stores frames as a sequence of image planes and pixel components (`channels`) can be spread out between these planes. Some formats use one plane and a pixel's components are packed/contigous (aka, the format is channel-last). For these, `n_channels` counts how many color channels the format has. Other formats store one component per plane (aka the format is channel-first) and for those formats `n_planes` counts the number of channels the format has.

Finally, there are esoteric formats. `n12` and consorts store a different number of of components per plane; this is not trivially strideable so we can't easily create a ndarray from this and I'm raising for this format. There are also `bayer_rggb8` and consorts, which are stridable, but pixels aren't packed so the notion of `channels` inside `shape` doesn't make sense anymore. I'm currently raising for this format as well, but I'm very open for suggestions on how to support it (because it is possible, but the question is how to do so cleanly).

File: imageio/plugins/pyav.py
Comment (FirefoxMetzger) [reply imageio/plugins/pyav.py:157]: Oh? Where is the other open?

File: imageio/plugins/pyav.py
Comment (almarklein) [reply imageio/plugins/pyav.py:90]: Maybe just put what you just said in a comment :)

File: imageio/plugins/pyav.py
Comment (almarklein) [reply imageio/plugins/pyav.py:157]: A few lines up where we establish whether we can read the file.

File: imageio/plugins/pyav.py
Comment (FirefoxMetzger) [reply imageio/plugins/pyav.py]: The auto-detection checks for the container's general ability to store variable framerate. This doesn't say anything about the actual timings of frames inside a file so we allow the user to control this if needed (because it is _really_ slow to seek inside of files with non-constant framerate).

File: imageio/plugins/pyav.py
Comment (FirefoxMetzger) [reply imageio/plugins/pyav.py:157]: Ah. The open here is for writing and the open call above for reading. Do you mean to try and combine the reading and writing path where possible?

File: imageio/plugins/pyav.py
Comment (FirefoxMetzger) [reply imageio/plugins/pyav.py:793]: Sure. This here is simply 1:1 plumbing. Inside FFmpeg these flags are called `key_frame` and `interlaced_frame`, so I figured it might be intuitive for users coming from that background to have the same names. Should we add a `is_` variant on top of this or as a replacement?

Comment (almarklein) [reply imageio/plugins/pyav.py:793]: If we do it, we should replace it. 

Good point - using the same terminology as ffmpeg is also worth something, so I'm fine if you leave it as-is.

File: tests/test_pyav.py
Comment (FirefoxMetzger) [reply tests/test_pyav.py]: @hmaarrfk Done. After discussion with @almarklein and with local colleagues I changed the default for reading in pyav to `rgb24`. Passing `format=None` will still read the native colorspace, but it will no longer be the default. I've added a test to make sure both scenarios are covered.

File: imageio/plugins/pyav.py
Comment (FirefoxMetzger) [reply imageio/plugins/pyav.py:18]: I created https://github.com/imageio/imageio/issues/763 to keep track of this. Since we are using `ArrayLike` in other places in the library this is a larger refactor and I will deal with this in a separate PR.

File: imageio/plugins/pyav.py
Comment (FirefoxMetzger) [reply imageio/plugins/pyav.py:90]: Done.

File: tests/test_pyav.py
Comment (FirefoxMetzger) [reply tests/test_pyav.py:94]: done.

File: imageio/plugins/pyav.py
Comment (FirefoxMetzger) [reply imageio/plugins/pyav.py:804]: Then I will keep them as is, because my general understanding is to keep the `metadata` dict keys close to the names in the format/backend where possible.

**Comment by FirefoxMetzger** [comment]
Nice! I just sorted out the pending suggestions from your reviews @almarklein , @hmaarrfk (thanks for taking the time!). I also sorted out the astropy failures on pypy [by no longer installing astropy, because I can't work out where/why it fails and I also can't reproduce the failure locally].

What's left?

**Comment by FirefoxMetzger** [comment]
@daveisfera I can"t think of anything specific from the top of my head. The main thing that is missing is approval from @almarklein and - perhaps - some documentation, which I will look into after I finished https://github.com/imageio/imageio/pull/769).

If you want, you could help to proofread the docs and see if they are intelligible and user friendly, or - depending on your level of expertise, time, and interest - you can also review the PR itself and see if there is anything that you think should be added or changed.

**Review by almarklein — APPROVED** [review]

**Comment by almarklein** [comment]
LGTM, but CI does not approve yet?

**Comment by FirefoxMetzger** [comment]
One of the things it will trip over is the changes in pillow 9.1 which break our CI. We've merged the PR that fixes this already, but I still have to rebase this PR on them. Will do soon(TM).

**Comment by FirefoxMetzger** [comment]
@almarklein @daveisfera @hmaarrfk I added some narrative documentation for the pyAV plugin. If one of you has the time, could you have a look at it and check for legibility and grammar? Once that is done, we can merge the PR :)

File: imageio/plugins/pyav.py
Comment (almarklein) [inline imageio/plugins/pyav.py:5]:
```suggestion
This plugin wraps pyAV, a pythonic binding for the FFMPEG library.
```

File: imageio/plugins/pyav.py
Comment (almarklein) [inline imageio/plugins/pyav.py:7]:
```suggestion
It is similar to our FFMPEG plugin, but offers a more performant and
robust interface, and aims to supersede the FFMPEG plugin in the future.
```

File: imageio/plugins/pyav.py
Comment (almarklein) [inline imageio/plugins/pyav.py:74]: This is cool, but also leaning towards video processing, which is not our core business. We should be careful not to "pollute" the docs too much. I guess it's ok here though, since it's the last section in theses docs.

File: imageio/plugins/pyav.py
Comment (almarklein) [inline imageio/plugins/pyav.py:536]: Maybe we should go easy on the type hints (e.g. using a string instead?), because this is not helping 😮  :

image

**Review by almarklein — APPROVED** [review]

File: imageio/plugins/pyav.py
Comment (almarklein) [reply imageio/plugins/pyav.py:74]: Are there docs in pyav that we can delegate some of this to? Perhaps some of this would fit as an actual example, instead of a code sample in the docs.

Comment (FirefoxMetzger) [reply imageio/plugins/pyav.py:74]: Unfortunately, no. Here are the docs on filters in pyav: https://pyav.org/docs/stable/api/filter.html . I worked out how to use them in pyav by digging through GH issues and source code. An item for the example page could, indeed, be nice, and I was already thinking of turning our GIF writing unit test into one. 

The downside of doing just an example is that the examples page is supposed to give various (general) usage examples of ImageIO, whereas the page here is supposed to discuss the features of the pyav plugin. The problem with that is that discussing/describing filter graphs is cumbersome, and a single example does more explaining than me writing a wall of text. That's essentially why I choose to put an example here. We could move them, but then we would need a decent (and clear) replacement for them here, because in the end we have to document available kwargs.

**Comment by daveisfera** [comment]
> @almarklein @daveisfera @hmaarrfk I added some narrative documentation for the pyAV plugin. If one of you has the time, could you have a look at it and check for legibility and grammar? Once that is done, we can merge the PR :)

Looked good to me. I'm not totally sure how I'd use this when [iterating over a movie](https://imageio--745.org.readthedocs.build/en/745/examples.html?highlight=iter#iterate-over-frames-in-a-movie) (is it now the default or do I need to do something to use this new code?), but that's a detail and I don't have any concerns with these changes.

Looking forward to being able to use this!

**Comment by FirefoxMetzger** [comment]
> I'm not totally sure how I'd use this when [iterating over a movie](https://imageio--745.org.readthedocs.build/en/745/examples.html?highlight=iter#iterate-over-frames-in-a-movie) (is it now the default or do I need to do something to use this new code?)

If I understand the question correctly then the answer is that, for now, you will have to explicitly select this plugin using the `plugin=` kwarg, e.g. via

```python
import imageio.v3 as iio

for frame in iio.imiter("imageio:cockatoo.mp4", plugin="pyav"):
    pass  # do stuff
```

because `ìmageio-ffmpeg` will remain the default for MP4. That is until this plugin had some time to prove itself and we have verified that it is indeed stable, feature-complete, and efficient enough to warrant switching the default.

File: imageio/plugins/pyav.py
Comment (almarklein) [reply imageio/plugins/pyav.py:74]: Mmm, I guess we can then best leave it like this for now.

**Comment by FirefoxMetzger** [comment]
Alright, looks like this PR is ready to go. I will wait with merging it until #779 is ready and then I will merge both in the same week.

This way, they will both be released the same week and we can advertise this plugin with the plugin-specific default for `index` (which is to read all frames/images by default instead of just the first). Should be less confusing to users than to experience this plugin switching behaviors.

**Comment by FirefoxMetzger** [comment]
The time has finally come. Once CI is happy, I will merge the pyav plugin, Thanks @almarklein @hmaarrfk @daveisfera for your feedback and discussion.

**Pull request merged**
